### PR TITLE
Add reversed actions when using reverse

### PR DESCRIPTION
**Pull request by halfmatthalfcat**

### CODE REVIEW

**Comment by halfmatthalfcat** [comment]
Looks like this is breaking a ton of stuff...hmmmm

**Comment by halfmatthalfcat** [comment]
So it looks like reverse is reversing within an operation (as evident by the tests) but it's not reversing when chaining operations

**Comment by nafg** [comment]
Can you confirm that your test fails without the change and passes with it?

**Comment by halfmatthalfcat** [comment]
@nafg I think I actually identified the root cause as https://github.com/nafg/slick-migration-api/blob/master/src/main/scala/slick/migration/api/Dialect.scala#L184

Sorting by the Action sort value is forcing order regardless of the appended order of the actions

**Comment by nafg** [comment]
Thanks for investigating this. I had forgotten about `sort`, but it actually exists for this very purpose. So the solution is actually much simpler. The `sort` of things like `DropForeignKey` should be lower than things like `DropTable`. You should just be able to fiddle with the numbers in `object Action`. (I'd recommend reordering the lines as well, so that they are in increasing `sort` order, so that it's easy to see the order. That's how it is now as well, except that the `sort` and order are wrong as you've discovered.)

P.S. Why is GitHub showing a changed line in build.sbt?

**Comment by halfmatthalfcat** [comment]
The alignment of all the dependency versions were correct except for that one haha.

**Comment by halfmatthalfcat** [comment]
@nafg yeah so my proposed solution was to add a `reverse` to the `migrateTable` function and if you've reversed the table, it'll reverse the sort of the sql strings https://github.com/nafg/slick-migration-api/pull/87/files#diff-e597e55948ec9df1caeac6f1b01e9040R184

I'm using this locally right now and it works as expected. Does this solution seem good to you?

**Comment by halfmatthalfcat** [comment]
Ok @nafg this should be good if you're ok with the solution. It looks like on every db except Postgres, you can't create, add and then remove all columns (there must be at least one). Not sure if that should be noted anywhere but this now acts as expected.

**Comment by nafg** [comment]
Can you comment on this suggestion?

> I had forgotten about `sort`, but it actually exists for this very purpose. So the solution is actually much simpler. The `sort` of things like `DropForeignKey` should be lower than things like `DropTable`. You should just be able to fiddle with the numbers in `object Action`. (I'd recommend reordering the lines as well, so that they are in increasing `sort` order, so that it's easy to see the order. That's how it is now as well, except that the `sort` and order are wrong as you've discovered.)

**Comment by halfmatthalfcat** [comment]
@nafg Ah sorry I think I had mis-understood what you were originally talking about. I played around with the weights (sort values) and have verified locally it now works for me. Is this more of what you were thinking?

**Comment by nafg** [comment]
Yes, thanks.

Do you mind explaining the tests in more detail?

**Comment by halfmatthalfcat** [comment]
It tests that a successful reversal of multiple chained migrations works. All of the other test examples, afaik, only test the reversal of one migration. 

Due to db mechanics for the various kinds though, Postgres is the only db that can drop all columns. All the others throw and error stating that you must have at least one column present. 

This test ensures that at least Postgres can do a compete reversal, as the others throw and we just continue on. I’m open to removing it though since really the only thing we’re testing is Postgres succeeding.

**Comment by nafg** [comment]
Should .reverse just be smarter and if it's reversing a create table skip
reversing create columns?

**Comment by halfmatthalfcat** [comment]
We could detect a drop table and then just drop any drop column actions from the actions list...we probably could just drop all other actions since dropping the table is the most final action of all actions. I'm not sure that's the most prudent way to go about it though, seems somewhat brute-forcy. 

I'm using Postgres personally, so I'm unaffected by the last column rule.

**Comment by nafg** [comment]
I hear, I just don't like the idea of a test using isInstanceOf and
try/catch like that, it seems like it's much less clear that everything
works 100%.

Some things, like indexes, I'm not sure if dropping the table "includes"
dropping them -- might it need a CASCADE? Or am I thinking of something
else? Anyway they are separate, top-level objects, while columns are part
of a table. So while dropping an index might be necessary and implied by
dropping the table, it's still a distinct action. On the other hand,
dropping columns and then the table is just dropping part of a thing and
then the rest. Dropping columns isn't *implied* by dropping the table, it's
*included* in it.

And that's besides the fact that apparently outside of postgres dropping
all the columns is an *error*, while nothing else is.

So in short, for multiple reasons it's not crazy to "drop" the drop-columns
when there's a drop-table too.

What do you think?

**Comment by nafg** [comment]
Hi, do you plan to implement this?

**Comment by halfmatthalfcat** [comment]
@nafg I don't have a lot of time right not unfortunately. I found a solution I can live with in my project for now. If I find some time in the future, I'll be able to revisit this if it's not already fixed.

**Comment by anovstrup** [comment]
@nafg Hi Naftoli, I just ran into this issue as well.  If you can help me understand what's still needed to get a solution over the finish line, I'd be happy to work on it.

**Comment by nafg** [comment]
@anovstrup would you be interested in pair programming it with me?

IIRC it wouldn't be much work but I have to context-switch back into it

**Comment by nafg** [comment]
> Sure! I probably won't be available until late this week (Thursday or Friday).

Great, can you message me on Discord?

> Sounds like this is what's left to implement, plus any additional testing that's needed?

I would have to refresh my memory ;)
